### DISCUSSION TITLE
Fetcher and mutator types

### DISCUSSION THREAD

**Discussion by ver1fy** [vercel/swr#2214] [Q&A] [answered]
I have been experimenting with the new `useSWRMutation` hook in 2.0.0-rc.0 and appear to have found a limitation with TypeScript. Assume we have the following fetcher:

```
const createPost = (post: PostDTO) => axios.post<Post>('/posts', post).then((res) => res.data);
```

If successful, the endpoint returns the newly created `Post` resource. We might write a custom hook to update a collection of posts like so:

```
const { trigger } = useSWRMutation('/posts', (_, options) => createPost(options.arg), {
  populateCache: (result, currentData) => [result, ...currentData],
  revalidate: false,
})
```

The problem here is that `currentData` is typed to the fetcher response (`Post`) and cannot be used to update a collection of posts (`Post[]`). In other words, you can only update a single `Post` with a `Post`. Unless there are some type assertions that can be added, the only way I can see to get this working is to omit the fetcher type assertion.

Is there something I am doing wrong or perhaps completely misunderstood the purpose of `useSWRMutation`? [This section of the docs](https://swr-site-n71kwepxk.vercel.sh/docs/mutation#mutate-based-on-current-data) would suggest this is a perfectly valid use case. The difference here is that `useSWRMutation` is being used instead of `mutate`.

Any guidance is much appreciated!

**Comment by samuel-bluesight** [comment]
I've been sharing the same problem. @shuding as i've seen you talking about best practices, do you have any recommendations to handle that case?

**Reply by ver1fy** [reply, marked as answer]
It's been a while since I've used SWR as I felt mutations weren’t quite ready at the time. Looking at this now, it would appear that `currentData` is no longer typed to the fetcher response, but rather `any`. What this means is that TypeScript will allow you to aggregate the data like so: `[result, ...currentData]`.

According to the docs, you'll probably want something like this:
```
const createPost = (url: string, { arg }: { arg: Post }) => {
  return axios.post<Post>(url, arg).then(res => res.data);
}

const useCreatePost = () => useSWRMutation('/posts', createPost, {
  populateCache: (newPost, posts = []) => [newPost, ...posts],
  revalidate: false
});
```

Alternatively, if you prefer not to couple your API calls to SWR like me, you could wrap the fetcher like so:
```
const createPost = (post: PostDTO) => axios.post<Post>('/posts', post).then(res => res.data);

const useCreatePost = () => useSWRMutation('/posts', (_, { arg }: { arg: Post }) => createPost(arg), {
  populateCache: (newPost, posts = []) => [newPost, ...posts],
  revalidate: false
});
```
You could also use the `mutate` callback returned by `useSWR`:
```
mutate<Post[]>('/posts', async (posts = []) => {
  const newPost = await createPost(post);
  return [newPost, ...posts];
}, { revalidate: false })
```
Hope this helps!

**Reply by samuel-bluesight** [reply]
That makes sense. I've made a couple of tests with `useSWRMutation` and I think the main outcome is to have the `populateCache` arguments typed in some way, like you did initializing it with empty array `[]`. That works, thanks!!
